src/Interpreters/InDepthNodeVisitor.h

```diff
@@ -95,4 +95,33 @@ class OneTypeMatcher
 template <typename Data, NeedChild::Condition need_child = NeedChild::all>
 using ConstOneTypeMatcher = OneTypeMatcher<Data, need_child, const ASTPtr>;
 
+template <typename Visitor, typename T = ASTPtr>
+struct InDepthNodeVisitorWithChildInfo : Visitor
+{
+    using ChildInfo = typename Visitor::ChildInfo;
```

Comment:
Why do we need special visitor for this, can't we use regular one and store the `info` in the `data`?

Reply:
I want explicitly show that:
1. Node processing result depends on the results of children nodes.
2. Children are processed completely independently. `Data` represents global visitor values.

Reply:
For example, `GetAggregatesVisitor` could be implemented in mode efficient way using this approach.
Instead of traversing query AST for several times in `getAggregates` and `getWindowFunctions` functions, we can collect functions and check correctness on the fly if we know independent results for child nodes.